skimage/morphology/_rolling_ball_cy.pyx

```diff
@@ -0,0 +1,220 @@
+                 DTYPE_FLOAT[::1] kernel not None,
+                 DTYPE_FLOAT[::1] ellipsoid_intensity not None,
+                 Py_ssize_t[::1] img_shape not None,
+                 Py_ssize_t[::1] padded_img_shape not None,
+                 Py_ssize_t[::1] kernel_shape not None,
+                 Py_ssize_t num_threads=0):
+    """
+    apply_kernel(img, kernel, ellipsoid_intensity, img_shape, padded_img_shape,
+                 kernel_shape, num_threads)
+
+    Apply a ND kernel to an ND image.
+
+    This function is the critical piece of code for 
+    `morphology.rolling_ellipsoid`. It was moved to cython for speed.
+
+    Parameters
+    ----------
+    img : (I) ndarray
+        A flat view into a padded image, e.g., from ``numpy.ravel()``.
+    kernel : (K) ndarray
+        A flat view into the kernel, e.g., from ``numpy.ravel()``. Indicates
+        if a pixel lies inside the ellipsoid. `kernel[x,y] == 1` if the pixel
+        is inside, ``kernel[x,y] == np.Inf`` otherwise.
+    ellipsoid_intensity : (K) ndarray
+        A flat view into the intensity of the ellipsoid, e.g., from 
+        ``numpy.ravel()``. Indicates the height/intensity of the ellipsoid at
+        position ``(x,y)``.
+    img_shape : (N) ndarray
+        The shape of the unflattened, unpadded image
+    padded_img_shape : (N) ndarray
+        The shape of the unflattened, padded image
+    kernel_shape : (N) ndarray
+        The shape of the unflattened kernel
+
+    Returns
+    -------
+    out_data : ndarray
+        the estimated background intensity. ``out_data.shape = img_shape``.
+
+    See Also
+    --------
+    rolling_ellipsoid
+
+    Notes
+    -----
+    - This function assumes that the image doesn't contain ``NaN``s; this 
+    assumption allows for faster code (better compiler optimization).
+    """
+    
+    cdef DTYPE_FLOAT[::1] out_data = np.zeros(np.prod(img_shape.base),
+                                              dtype=img.base.dtype)
+    cdef Py_ssize_t offset, offset_idx, out_data_size, ker_idx, img_idx
+    cdef DTYPE_FLOAT min_value, tmp
+    out_data_size = out_data.size
+
+    cdef Py_ssize_t ndim = kernel_shape.shape[0]
+    cdef Py_ssize_t kernel_outer = np.prod(kernel_shape[0:(ndim - 1)])
+    cdef Py_ssize_t kernel_inner = kernel_shape[ndim - 1]
+    cdef Py_ssize_t ker_idx_outer, ker_idx_inner
+
+    for offset_idx in prange(
+            out_data_size,
+            num_threads=num_threads,
+            nogil=True):
+        offset = ind2ind(offset_idx, 0, img_shape, padded_img_shape)
+        min_value = INFINITY
+
+        # split into outer and inner loop for vectorization
+        # (the inner loop is contiguous in memory)
+        for ker_idx_outer in range(kernel_outer):
+            ker_idx_outer = ker_idx_outer * kernel_inner
+            img_idx = ind2ind(ker_idx_outer, offset, kernel_shape,
+                              padded_img_shape)
+            for ker_idx_inner in range(kernel_inner):
```

Comment:
Ok so in terms of my earlier confusion, I might suggest renaming kernel_outer and kernel_inner to kernel_row_size and kernel_num_rows. Although these do not describe exactly the shape in n-dimensions, they are more descriptive than "outer/inner", and they are correct in 2D. And anyway an alternative nomenclature is `kernel_run_size` and `kernel_num_runs`. Then, `kernel_run_idx`, `kernel_idx_in_run`, and `kernel_idx`. What do you think?

Reply:
I think this is contingent on the community maintaining the code later xD Inner/outer loop seems more common in the HPC community (see for example [here](https://software.intel.com/content/www/us/en/develop/articles/outer-loop-vectorization.html) for some super nerdy vectorization), but nrow/ncol or num_rows/row_size might make more sense when coming from a image processing background. I am worried that the "not correct in ND" part will cause confusion later since this funny looping exists specifically to handle the ND case.

Perhaps something like `other_dim_size` and `last_dim_size`? I fear that won't be much better though.

Reply:
Renamed to `kernel_leading_dims` (former: kernel_outer) and `kernel_last_dim` (former: kernel_inner)

Reply:
I've also added a comment to both inner loops stating that they were separated for speed (vectorization)